File: nodejs/src/motoko-helper.ts
import {init, WASI} from '@wasmer/wasi';

export enum MoTag {
    BLOB = 17,
    CONCAT = 25,
}

export interface MoObj {
    tag: MoTag;
}

export interface MoBlob {
    tag: MoTag;
    len: number;
}

export interface MoConcat {
    tag: MoTag;
    bytes: number;
    text1: number;
    text2: number;
}

const utf8Decoder = new TextDecoder("utf8");
const utf8Encoder = new TextEncoder();

function toObj(
    arr: Uint32Array
): MoObj {
    return {
        tag: arr[0],
    };
}

function toBlob(
    arr: Uint32Array
): MoBlob {
    return {
        tag: arr[0],
        len: arr[1],
    };
}

function toConcat(
    arr: Uint32Array
): MoConcat {
    return {
        tag: arr[0],
        bytes: arr[1],
        text1: arr[2],
        text2: arr[3],
    };
}

export class MoHelper {
    wasi?: WASI;
    module?: WebAssembly.Module;
    instance?: WebAssembly.Instance;
    memory: WebAssembly.Memory = {} as any;
    allocBlob: (bytes: number) => number = () => 0;
    
    constructor() {
    }

    async load(
        wasm: Buffer
    ): Promise<boolean> {
        await init();

        this.wasi = new WASI({});
    
        this.module = await WebAssembly.compile(wasm);

        return !!this.module;
    }

    instanciate(
    ): boolean {
        this.instance = this.wasi?.instantiate(this.module, {});
        if(!this.instance) {
            return false;
        }
        this.wasi?.start(this.instance);
        this.memory = this.instance.exports.memory as any;
        this.allocBlob = this.instance?.exports.alloc_blob as any;
        return true;
    }

    call(name: string, ...args: any[]): any {
        const func = this.instance?.exports[name] as any;
        return func(...args);
    }

    textToString(
        skewedPtr: number
    ): string {
        const ptr = skewedPtr + 1;
        const arr = new Uint32Array(this.memory.buffer, ptr);
        const obj = toObj(arr);
        switch(obj.tag) {
            case MoTag.BLOB:
                const blob = toBlob(arr);
                const bytes = new Uint8Array(this.memory.buffer, ptr + 8, blob.len);
                return utf8Decoder.decode(bytes);
            case MoTag.CONCAT:
                const concat = toConcat(arr);
                return this.textToString(concat.text1) + this.textToString(concat.text2);
            default:
                return "";
        }
    }

    stringToText(
        s: string
    ): number {
        const encoded = utf8Encoder.encode(s);
        
        const skewedPtr = this.allocBlob(encoded.length);
        const ptr = skewedPtr + 1;
        
        const bytes = new Uint8Array(this.memory.buffer, ptr + 8);
        for(let i = 0; i < encoded.length; i++) {
            bytes[i] = encoded[i];
        }
        
        return skewedPtr;
    }
}

File: flutter/lib/motoko_helper.dart
import 'dart:convert';

import 'package:flutter_wasm/flutter_wasm.dart';

enum MoTag {
  blob(17),
  concat(25);

  const MoTag(this.value);
  final int value;
  static MoTag getByValue(num i) {
    return MoTag.values.firstWhere((x) => x.value == i);
  }
}

class MoObj {
  MoTag tag;
  MoObj(this.tag);
}

class MoBlob {
  MoTag tag;
  int len;
  MoBlob(this.tag, this.len);
}

class MoConcat {
  MoTag tag;
  int bytes;
  int text1;
  int text2;
  MoConcat(this.tag, this.bytes, this.text1, this.text2);
}

int toUint32(int a, int b, int c, int d) {
  return a | b << 8 | c << 16 | d << 24;
}

MoObj toObj(List<int> list) {
  return MoObj(MoTag.getByValue(toUint32(list[0], list[1], list[2], list[3])));
}

MoBlob toBlob(List<int> list) {
  return MoBlob(MoTag.getByValue(toUint32(list[0], list[1], list[2], list[3])),
      toUint32(list[4], list[5], list[6], list[7]));
}

MoConcat toConcat(List<int> list) {
  return MoConcat(
    MoTag.getByValue(toUint32(list[0], list[1], list[2], list[3])),
    toUint32(list[4], list[5], list[6], list[7]),
    toUint32(list[8], list[9], list[10], list[11]),
    toUint32(list[12], list[13], list[14], list[15]),
  );
}

class MoHelper {
  WasmInstance instance;
  dynamic allocBlob;

  MoHelper(this.instance) : allocBlob = instance.lookupFunction("alloc_blob");

  String textToString(int skewedPtr) {
    var ptr = skewedPtr + 1;
    var arr = instance.memory.view.skip(ptr).toList();
    var obj = toObj(arr);
    switch (obj.tag) {
      case MoTag.blob:
        var blb = toBlob(arr);
        var bytes = arr.skip(8).take(blb.len).toList();
        return const Utf8Decoder().convert(bytes);
      case MoTag.concat:
        var conc = toConcat(arr);
        return textToString(conc.text1) + textToString(conc.text2);
      default:
        return '';
    }
  }

  int stringToText(String s) {
    var encoded = const Utf8Encoder().convert(s);
    var skewedPtr = allocBlob(encoded.lengthInBytes);
    var ptr = skewedPtr + 1;

    for (var i = 0; i < encoded.lengthInBytes; i++) {
      instance.memory.view[ptr + 8 + i] = encoded[i];
    }

    return skewedPtr;
  }
}
